tunnel/intra/doh/padding.go

```diff
@@ -0,0 +1,112 @@
+// Copyright 2019 The Outline Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package doh
+
+import (
+	"golang.org/x/net/dns/dnsmessage"
+)
+
+const (
+	OptResourcePaddingCode = 12
+	OptDefaultPaddingLen   = 400
+)
+
+const kOptRrHeaderLen int = 1 + // DOMAIN NAME
+	2 + // TYPE
+	2 + // CLASS
+	4 + // TTL
+	2 // RDLEN
+
+const kOptPaddingHeaderLen int = 2 + // OPTION-CODE
+	2 // OPTION-LENGTH
+
+// Compute the number of padding bytes needed, excluding
+// headers. Assumes that |msgLen| is the length of a raw DNS message
+// excluding any RFC7830 padding option.
+func computePaddingSize(msgLen int, hasOptRr bool, blockSize int) int {
+	var padSize int = 0
+	// We'll always be adding a new padding header inside the OPT
+	// RR's data.
+	var extraPadding = kOptPaddingHeaderLen
+
+	// If we don't already have an OPT RR, we'll need to add its
+	// header as well!
+	if !hasOptRr {
+		extraPadding += kOptRrHeaderLen
+	}
+
+	for (msgLen+extraPadding+padSize)%blockSize != 0 {
```

Comment:
I think this can be `return blockSize-(msgLen+extraPadding+padSize)%blockSize`

Reply:
OK, there are a few more edge cases. Let me know what you think of the new version.